### PR TITLE
Fixing the issue of selection order by won't give correct ordering

### PR DESCRIPTION
**Pull request by fx19880617**
Current behavior:
Selection Order by query will always order the results based on the selection list sequence not the columns provided.

E.g. for query:
```
SELECT a,b,c FROM mytable ORDER BY c
```
The results will be ordered by a not c.

Sample bad queries from quickstart:
```
select AtBatting,G_old from baseballStats order by G_old desc ,hits desc limit 10
```

### CODE REVIEW

**Comment by fx19880617** [comment]
I don't think we move column sequence now. Server data table is built based on selection column list

**Comment by Jackie-Jiang** [comment]
@fx19880617 Check SelectionOperatorUtil.extractDataSchema, sort columns are put in front of the other columns. But seems @kishoreg recently re-write the SelectionOperator and that code is not invoked. I am not sure about the current behavior.

**Comment by fx19880617** [comment]
Server side will return a data table with schema. broker side reduce logic should be agnostic to the ordering of schema from server side.

**Comment by Jackie-Jiang** [comment]
If server side guarantees that the sorted columns are always put in front of other columns, then broker side does not need this extra index (where the value for each index is just the index) which will add overhead.
I tried your new test on the existing code and it passed, so I don't understand what is the purpose of this fix.

**Comment by fx19880617** [comment]
Current behavior keeps the column ordering from selection list and not put ordered columns in front.

**Comment by fx19880617** [comment]
The selection test query comparison logic only check the record existence, so it won't recognize the disordered behavior. You can run the sample query in PR description in quickstart to see the behavior.

**Comment by Jackie-Jiang** [comment]
In that case, can you please make the selection test query comparison check the order? With that we can prevent breaking the order in the future.

File: pinot-core/src/main/java/org/apache/pinot/core/query/selection/SelectionOperatorUtils.java [hunk 1]
```diff
@@ -67,14 +67,13 @@
...
 public class SelectionOperatorUtils {
-  private SelectionOperatorUtils() {
-  }
-
   private static final String INT_PATTERN = "##########";
   private static final String LONG_PATTERN = "####################";
   private static final String FLOAT_PATTERN = "#########0.0####";
   private static final String DOUBLE_PATTERN = "###################0.0#########";
   private static final DecimalFormatSymbols DECIMAL_FORMAT_SYMBOLS = DecimalFormatSymbols.getInstance(Locale.US);
+  private SelectionOperatorUtils() {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/query/selection/SelectionOperatorUtils.java]: (nit) let's not move this as all the utils classes put the private constructor at the top

File: pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/ClusterIntegrationTestUtils.java [hunk 2]
```diff
@@ -766,11 +772,25 @@ public static void testQuery(@Nonnull String pinotQuery, @Nonnull String queryFo
           }
           String actualValue = actualValueBuilder.toString();
 
-          // Check actual value in expected values set
-          if (!expectedValues.contains(actualValue)) {
-            String failureMessage = "Selection result returned in Pinot but not in H2: " + actualValue;
-            failure(pinotQuery, sqlQueries, failureMessage);
-            return;
+          final BrokerRequest brokerRequest = PinotQueryParserFactory.get("pql").compileToBrokerRequest(pinotQuery);
```
Comment (Jackie-Jiang) [inline pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/ClusterIntegrationTestUtils.java]: (nit) We don't usually put final in method

File: pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/ClusterIntegrationTestUtils.java [hunk 3]
```diff
@@ -766,11 +772,25 @@ public static void testQuery(@Nonnull String pinotQuery, @Nonnull String queryFo
...
+          final List<String> selectionColumns = brokerRequest.getSelections().getSelectionColumns();
+          final Set<String> orderByColumns = getOrderByColumns(brokerRequest.getOrderBy());
```
Comment (Jackie-Jiang) [inline pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/ClusterIntegrationTestUtils.java]: Please use `brokerRequest.getSelections().getSelectionSortSequence()` as the orderByColumns

File: pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/ClusterIntegrationTestUtils.java [hunk 4]
```diff
@@ -766,11 +772,25 @@ public static void testQuery(@Nonnull String pinotQuery, @Nonnull String queryFo
...
+
+          final List<SelectionSort> orderBy = brokerRequest.getOrderBy();
+          if ((orderBy != null) && (!orderBy.isEmpty()) && selectionColumns.containsAll(orderByColumns))  {
```
Comment (Jackie-Jiang) [inline pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/ClusterIntegrationTestUtils.java]: Why do we need the selectionColumns.containsAll(orderByColumns)?

File: pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/ClusterIntegrationTestUtils.java [hunk 5]
```diff
@@ -766,11 +772,25 @@ public static void testQuery(@Nonnull String pinotQuery, @Nonnull String queryFo
...
+            // Check actual value in expected values set
+            if (!expectedValues.get(rowIndex).equals(actualValue)) {
```
Comment (Jackie-Jiang) [inline pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/ClusterIntegrationTestUtils.java]: We need to handle the tie case

Comment (fx19880617) [reply pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/ClusterIntegrationTestUtils.java]: it's impossible from the results. E.g. `select a from table, order by b,c`, if the tie results more than limit size, then we don't event know the ordering.

The only thing we can determine this is that order by columns are the same as selection columns.

**Pull request merged**